Add support for fargate launch_type in ecs_service module and ecs_taskdefinition module

mjmayer opened the pull request:
##### SUMMARY
Add support for fargate launch_type in ecs_service module and ecs_taskdefinition module. This fixes #35607 . There are now two types of ECS clusters in AWS, EC2 and Fargate ,this adds support for Fargate.

Added some additional error checking to make sure when the playbook is executed it identifies required values such as memory, cpu, and network_type, and network_configuration before reaching out to the AWS API.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ecs_service, ecs_taskdefinition

##### ANSIBLE VERSION
```
ansible 2.6.0dev0 (devel eb9fe31e51) last updated 2018/05/25 14:26:56 (GMT -700)
  config file = None
  configured module search path = [u'/home/mjmayer/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mjmayer/ansible_mjmayer/lib/ansible
  executable location = /home/mjmayer/ansible_mjmayer/bin/ansible
  python version = 2.7.5 (default, Apr 11 2018, 17:41:36) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28.0.1)]

```

jsteinich committed:
add launchType to ecs_service per #35607

update ecs_taskdefinition to support fargate tasks per #35607

mjmayer committed:
Add Fargate support for ECS modules

Fargate instances do not require memory and cpu descriptors. EC2 instances
 do require descriptions. https://botocore.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.describe_task_definition

Fargate requires that cpu and memory be defined at task definition level.
EC2 launch requires them to be defined at the container level.

Fargate requires the use of awsvpc for the networking_mode. Also updated,
the documentation regarding where and when memory/cpu needs to the assigned.

The task_definition variable for the awspvc configuration colided with
the ecs_service for the bridge network. This would cause the test to fail.

Add testing for fargate

Add examples for fargate and ec2

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -256,6 +288,17 @@ def main():
             module.fail_json(msg="To use task definitions, a family must be specified")
 
+        launch_type = module.params['launch_type']
+        if launch_type == 'FARGATE' and ('cpu' not in module.params or not module.params['cpu']):
```
These could be required_if in AnsibleModule

```
required_if(['launch_type', 'FARGATE', ['cpu', 'memory']])
```

You'll need your existing network_mode check

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -174,12 +193,21 @@ def register_task(self, family, task_role_arn, network_mode, container_definitio
+            networkMode=network_mode,
+            containerDefinitions=container_definitions,
+            volumes=volumes,
+            requiresCompatibilities=[launch_type]
```
This doesn't seem to have any checks for whether boto3 supports `requiresCompatibilities`, which seems to only come with version 1.8.4 and beyond of botocore.

You'll need a check similar to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L445

`requiresCompatibilities` should only be added to the parameter list if it's actually required.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -368,7 +376,8 @@ def create_service(self, service_name, cluster_name, task_definition, load_balan
             placementConstraints=placement_constraints,
-            placementStrategy=placement_strategy)
+            placementStrategy=placement_strategy,
+            launchType=launch_type)
```
This doesn't seem to have any checks for whether boto3 supports `launchType`, which seems to only come with version 1.8.4 and beyond of botocore.

You'll need a check similar to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L445

`launchType` should only be passed here if it's actually required.

mjmayer commented:
@willthames I've made the requested changes and update this PR.

willthames commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -414,6 +425,12 @@ def ecs_api_handles_network_configuration(self):
+    def ecs_api_supports_launch_type(self):
+        from distutils.version import LooseVersion
+        # Checking to make sure botocore is greater than a specific version.
+        # Support for placementStrategy is only available in versions beyond 1.8.4
```
This should read launchType, not placementStrategy

willthames commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -431,7 +448,8 @@ def main():
         placement_strategy=dict(required=False, default=[], type='list'),
-        network_configuration=dict(required=False, type='dict')
+        network_configuration=dict(required=False, type='dict'),
+        launch_type=dict(required=False, choices=['EC2', 'FARGATE'], default='EC2')
```
don't set a default here, or this won't work with `< 1.8.4`

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -234,9 +269,16 @@ def main():
         task_role_arn=dict(required=False, default='', type='str'),
-        volumes=dict(required=False, type='list')))
+        volumes=dict(required=False, type='list'),
+        launch_type=dict(required=False, choices=['EC2', 'FARGATE'], default='EC2'),
```
Don't set a default for `launch_type`

willthames commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -234,9 +269,16 @@ def main():
-        volumes=dict(required=False, type='list')))
+        volumes=dict(required=False, type='list'),
+        launch_type=dict(required=False, choices=['EC2', 'FARGATE'], default='EC2'),
+        cpu=dict(required=False, type='str'),
```
This can just be `cpu=dict()` as both `required=False` and `type='str'` are the default.

willthames reviewed (changes requested):
The main blocking issue now is just the default for `launch_type`.

However, that does identify for me that there are no tests associated with this change. 

The test for < 1.8.4 can be added to test/integration/targets/ecs_cluster/playbooks/network_fail.yml

The test for launch_type=fargate should be added to test/integration/targets/ecs_cluster/playbooks/full_test.yml

Let me know if you have any problems with the tests, and I'll see if I can help you resolve them.

mjmayer commented:
@willthames I have the ansible-test environment setup according to your [directions](http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html). I am running into failures of the tests for a previous PR #34381 
```
TASK [create ecs_service using awsvpc network_configuration] ******************************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.errorfactory.InvalidParameterException: An error occurred (InvalidParameterException) when calling the CreateService operation: Network Configuration is not valid for the given networkMode of this task definition.
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 584, in <module>\n    main()\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 532, in main\n    module.params['launch_type'])\n  File \"/tmp/ansible_vurQzX/ansible_module_ecs_service.py\", line 385, in create_service\n    response = self.ecs.create_service(**params)\n  File \"/home/vagrant/.local/lib/python2.7/site-packages/botocore/client.py\", line 314, in _api_call\n    return self._make_api_call(operation_name, kwargs)\n  File \"/home/vagrant/.local/lib/python2.7/site-packages/botocore/client.py\", line 612, in _make_api_call\n    raise error_class(parsed_response, operation_name)\nbotocore.errorfactory.InvalidParameterException: An error occurred (InvalidParameterException) when calling the CreateService operation: Network Configuration is not valid for the given networkMode of this task definition.\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
...ignoring

TASK [check that graceful failure message is returned from ecs_service] *******************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "assertion": "ecs_service_creation_vpc.msg == \"botocore needs to be version 1.7.44 or higher to use network configuration\"",
    "changed": false,
    "evaluated_to": false
}
```
It seems like the module.fail_json() is not actually stopping the execution of the module.
https://github.com/ansible/ansible/blob/f87ca27aa3dc5a017374f2204ede1568dbcb4515/lib/ansible/modules/cloud/amazon/ecs_service.py#L443-L448

willthames commented:
@mjmayer - thanks for addressing my feedback. I'll try and run these tests independently shortly.

I assume [the April 26 blog post](https://aws.amazon.com/about-aws/whats-new/2018/04/aws-fargate-now-available-in-ohio--oregon--and-ireland-regions/) is the latest fargate availability statement - what happens if the module is run outside of those regions?

willthames commented:
I've added two commits to your branch to improve the region handling so we don't have to maintain image ids for multiple regions forever, and so that updating compute-policy works again.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -458,10 +475,19 @@ def main():
         matching = False
         update = False
+
+        if module.params['launch_type'] == 'FARGATE' and ('network_configuration' not in module.params or not module.params['network_configuration']):
```
I think this can just be added to the require_if on [line 439](https://github.com/ansible/ansible/pull/40736/files#diff-5ab98c93ae8d77e786cf0afa8a0e95c3L439) with `('launch_type', 'FARGATE', ['network_configuration'])`

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_service.py:
```diff
@@ -458,10 +475,19 @@ def main():
     results = dict(changed=False)
+
+    if module.params['launch_type']:
+        if not service_mgr.ecs_api_supports_launch_type():
```
Rather than importing LooseVersion in the module you could use `if not module.botocore_at_least('1.8.4'):` (added a few days ago in 7f8654d586be176a355a685c249eb28fd5a17d0d)

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -234,9 +268,16 @@ def main():
+        volumes=dict(required=False, type='list'),
+        launch_type=dict(required=False, choices=['EC2', 'FARGATE']),
+        cpu=dict(),
+        memory=dict(required=False, type='str')
```
~Does this need to be an int? The docs https://boto3.readthedocs.io/en/latest/reference/services/ecs.html#ECS.Client.register_task_definition indicate that~
Oops, wrong parameter. Too many 'memory' options, ignore that.

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -73,6 +73,24 @@
+        version_added: 2.7
+    memory:
+        description:
+            - The amount (in MiB) of memory used by the task. If using the EC2 launch type, this field is optional and any value can be used.
```
It looks like this also accepts things like `'1GB'`

Do you mind adding an example or two to the docs?

s-hertel commented on lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py:
```diff
@@ -245,6 +286,10 @@ def main():
     results = dict(changed=False)
 
+    if module.params['launch_type']:
+        if not task_mgr.ecs_api_supports_requirescompatibilities():
```
This could also use `module.botocore_at_least()`

mjmayer replied:
It looks like the changes in 7f8654d where only for AnsibleAWSModule and ecs_taskdefinition is using AnsibleModule https://github.com/ansible/ansible/blob/6ef2ffe310b725b4f445fc0d95fff9954df94739/lib/ansible/modules/cloud/amazon/ecs_taskdefinition.py#L239

I saw this when running a test
```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'AnsibleModule' object has no attribute 'botocore_at_least'
```

willthames replied:
We should move ecs_taskdefinition to AnsibleAWSModule but we can do that in a separate PR :)

willthames commented:
Where do the first two commits come from - normally we'd just use github's squash and merge but that would obliterate @jsteinich's contributions (I'm not too worried about my trivial fixes)

mjmayer commented:
@jsteinich contributions can be found in https://github.com/PerBlue/ansible/tree/ecs_service_launch_type

jsteinich commented:
I apologize for not making a pull request for my changes. I meant to but was waiting for a dependent PR to be merged.

Let me know if you need anything. I actually ended up making a couple more commits outside of the branch to fix some things I found in testing: https://github.com/PerBlue/ansible/commit/417f1fcb2c0c5d673d6075a47f3f426f0427fc58
and
https://github.com/PerBlue/ansible/commit/9b32ffdebada75f0fb55b2be91e23ac7a8e4c966

willthames commented:
I'm happy for this to be merged - the remaining question is how best to do that while retaining attribution. I've got a version of your branch that squashes the whole lot into three commits to reflect the various authorships but I don't want to just force push that to your branch without asking.

willthames committed:
Remove ECS policies from AWS compute policy

The compute policy was exceeding maximum size and contained
policies that already exist in ecs-policy.

Look up suitable AMIs rather than hardcode

We don't want to maintain multiple image IDs for multiple regions
so use ec2_ami_facts to set a suitable image ID

Improve exception handling

willthames commented:
This is passing local tests - I'll merge once shippable passes.

willthames commented:
Merged, thanks @mjmayer and @jsteinich
